File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new LinkedList<>();
        if(root == null) return result;
        Queue<Queue<TreeNode>> BST = new LinkedList<>();
        Queue<TreeNode> initialList = new LinkedList<>();
        initialList.add(root);
        BST.add(initialList);
        while(!BST.isEmpty()) {
            Queue<TreeNode> levelStack = BST.poll();
            List<Integer> levelList = new LinkedList<>();
            Queue<TreeNode> nextLevel = new LinkedList<>(); 
            while(!levelStack.isEmpty()) {
                TreeNode node = levelStack.poll();
                levelList.add(node.val); //Add the node to the level list, in the proper order (as they are in the stack). Hopefully this is ok?
                if(node.left != null) nextLevel.add(node.left); //Push in proper order, that way BST orders correctly and handles single-child nodes.
                if(node.right != null) nextLevel.add(node.right);
            }
            if(!levelList.isEmpty()) result.add(levelList);
            if(!nextLevel.isEmpty()) BST.add(nextLevel);
        }
        //Will this work? Well, for each level, we first check out the stack. This pops the stack with just 3 in it, then adds 20, 9 in LIFO order.
        //Then the stack is empty, list with 3 is added, and stack with 20, 9 is added.
        //9 is popped. No childre, but a new stack and list is created, with the 9 inside the list.
        //20 is popped. 2 children, 7, 15 are added to stack in LIFO order.
        //Stack is now empty, so the new layer stack is added to the list, and the list is appended to.
        //What if 9 had a child? Well, it would be added to the stack, then 15 and 7. Oof, that won't work. I guess we do really need FIFO then?
        //And we need to reverse the order of the nodes.
        return result;
    }
}

File: 102-binary-tree-level-order-traversal/102-binary-tree-level-order-traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

/*
    (Slow) Recursive Method
    
            val layerList = mutableListOf<List<Int>>(listOf(root.`val`))
        var leftList = emptyList<List<Int>>()
        var rightList = emptyList<List<Int>>()
        if(root.left != null) {
            leftList = levelOrderAddToList(root.left)
        }
        if(root.right != null) {
            rightList = levelOrderAddToList(root.right)
        }
        if(!leftList.isEmpty()) {
            if(!rightList.isEmpty()) {
                //Neither Empty
                if(leftList.size > rightList.size) {
                    for(index in 0..leftList.size - 1) {
                        val list : List<Int> = leftList[index]
                        rightList.getOrNull(index)?.let { right ->
                            layerList.add(list + right)
                        } ?: layerList.add(list)
                    }
                } else {
                    for(index in 0..rightList.size - 1) {
                        val list : List<Int> = rightList[index]
                        leftList.getOrNull(index)?.let {left -> 
                            layerList.add(left + list)
                        } ?: layerList.add(list)
                    }
                }
            } else {
                //Right Empty
                leftList.forEach { layer -> layerList.add(layer) }
            }
        } else if (!rightList.isEmpty()) { //Left Empty
            rightList.forEach { layer -> layerList.add(layer) }
        } //Don't do anything if both empty
        return layerList
*/

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        //Sounds like BFS! But on a tree, instead of...well a graph
        /*
            Reminder: BFS is not a recursive function like DFS! You use a queue.
            
            Approach
            
                Start at the root
                For each layer, create a new list (or queue)
                For the current list that you have, go from left to right, and add children in the order in which they appear
                Once you're done, add the list to the queue of lists
                
            If n is the number of nodes in the tree, then arguably, O(n)
        */
        if(root == null) return listOf<List<Int>>()
        
        val returnList = mutableListOf<List<Int>>()
        val traversalList = mutableListOf<List<TreeNode>>(listOf(root))
        
        while(!traversalList.isEmpty()) {
            val front = traversalList.first()
            val nextLayerList = mutableListOf<TreeNode>()
            front.forEach { node -> 
                if(node.left != null) nextLayerList.add(node.left)
                if(node.right != null) nextLayerList.add(node.right)
            }
            if(!nextLayerList.isEmpty()) traversalList.add(nextLayerList)
            returnList.add(front.map{ node -> node.`val` })
            traversalList.removeAt(0)
        }
        return returnList
    }
}